src/core/service.c

```diff
@@ -1440,6 +1440,119 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static int service_copy_exit_status_env(char **env, size_t *n_env, Service *s) {
+        assert(s);
+        assert(env);
+        assert(n_env);
+
+        /* Copy service exit metadata from 's' to env. */
+
+        int r = asprintf(env + *n_env, "SERVICE_RESULT=%s", service_result_to_string(s->result));
+        if (r < 0)
+                return -ENOMEM;
+
+        (*n_env)++;
+
+        if (s->main_exec_status.pid > 0 && dual_timestamp_is_set(&s->main_exec_status.exit_timestamp)) {
+                r = asprintf(env + *n_env, "EXIT_CODE=%s", sigchld_code_to_string(s->main_exec_status.code));
+                if (r < 0)
+                        return -ENOMEM;
+        }
+
+        (*n_env)++;
+
+        if (s->main_exec_status.code == CLD_EXITED)
+                r = asprintf(env + *n_env, "EXIT_STATUS=%i", s->main_exec_status.status);
+        else
+                r = asprintf(env + *n_env, "EXIT_STATUS=%s", signal_to_string(s->main_exec_status.status));
+
+        if (r < 0)
+                return -ENOMEM;
+
+        (*n_env)++;
+
+        return 0;
+}
+
+static int service_copy_monitor_md_env (char **env, Job *j) {
+        assert(j);
+        assert(env);
+
+        bool first = true;
+        Unit *tu;
+
+        /* Add a single environment variable 'MONITOR_METADATA' to the environment.
+         * This variable contains a space separated set of fields which relate to
+         * the service(s) which triggered job 'j'. Job 'j' is the JOB_START job for
+         * an OnFailure= or OnSuccess= dependency. Format of the MONITOR_METADATA
+         * variable is as follows:
+         *
+         * MONITOR_METADATA="SERVICE_RESULT:<result-string0>,EXIT_CODE:<exit-code0>:EXIT_STATUS:<exit-status0>:INVOCATION_ID:<id>,UNIT:<triggering-unit0.service> SERVICE_RESULT:<result-stringN>,EXIT_CODE:<exit-codeN>:EXIT_STATUS:<exit-statusN>:INVOCATION_ID:<id>,UNIT:<triggering-unitN.service>"
```

Comment:
for parsing purposes, have you double checked that the result string/exit status cannot contain spaces/colons?

Reply:
Yep, we are good according to the definition of these values in the man page. However I'm not massively convinced about this pattern as I think it's a bit messy.

I'm thinking I'll replace the colons with equals instead as this would be clearer imo, I'll also replace the space with a semi-colon. So for example this would change to:

```MONITOR_METADATA="SERVICE_RESULT=<result-string0>,EXIT_CODE=<exit-code0>,EXIT_STATUS=<exit-status0>=INVOCATION_ID=<id>,UNIT=<triggering-unit0.service>;SERVICE_RESULT=<result-stringN>,EXIT_CODE=<exit-codeN>=EXIT_STATUS=<exit-statusN>=INVOCATION_ID=<id>,UNIT=<triggering-unitN.service>"```

Reply:
yeah, that looks better